File: scala/src/main/scala/domain/Portfolio.scala
package domain

case class Portfolio(moneys: Money*) {
  def evaluate(bank: Bank, toCurrency: String): Either[String, Money] = {
    val convertedMoneys = moneys.map(bank.convert(_, toCurrency))
    val lefts = convertedMoneys.collect { case Left(x) => x }

    if (lefts.isEmpty)
      Right(
        Money(
          convertedMoneys
            .collect { case Right(x) => x }
            .foldLeft(0d)((acc, money) => acc + money.amount),
          toCurrency
        )
      )
    else
      Left(lefts.mkString("Missing exchange rate(s): [", ",", "]"))
  }
}

File: c#/money-problem.Domain.C#/Portfolio.cs
using LanguageExt;
using static LanguageExt.Prelude;

namespace money_problem.Domain
{
    public record Portfolio(params Money[] Moneys)
    {
        public Either<string, Money> Evaluate(
            Bank bank,
            Currency toCurrency)
        {
            var convertedMoneys = Moneys
                .Map(m => bank.Convert(m, toCurrency))
                .ToList();

            return !convertedMoneys.Lefts().Any()
                ? new Money(convertedMoneys.Rights().Fold(0d, (acc, money) => acc + money.Amount), toCurrency)
                : Left($"Missing exchange rate(s): [{string.Join(",", convertedMoneys.Lefts())}]");
        }
    }
}